Add a few convenience functions to the utils module

ddasilva opened the pull request:
These are convenience functions, for debugging purposes by both users and package developers. 

* utils.read_primary_headers()
* utils.iter_packet_bytes()
* utils.split_packet_bytes()
* utils.count_packets()
* utils.get_packet_apid() 
* utils.get_packet_total_bytes()

ddasilva commented:
Doc build for this branch:
https://ccsdspy.readthedocs.io/en/added_utils/api.html

ehsteve commented:
These functions seem like they would be useful. They could really use some examples though because it is not very clear exactly what they do. For `count_packets`, how about issuing a warning if the last packet is not complete? Also some of the code is being repeated from other areas in the code base? Would it be possible to reduce duplication? Final question, do these functions work on both fixed length and variable length packets?

ddasilva commented:
Yes, these functions work with mixed APID streams which can include both variable length and fixed length packets. I added something about this in the function doc now to make this clear. 

I also wrote a user guide page with examples: 
https://ccsdspy.readthedocs.io/en/added_utils/user-guide/utils.html

I reduced a bit of code duplication in utils.py. Right now `count_packets()` uses its own loop because it gives the option of returning the number of missing bytes at the end (I want a way to calculate this programmatically without having to intercept a warning message).  I'm not sure how to bolt on that extra return value onto `iter_packet_bytes()`, which is a generator. 

I started down having the variable length decoder (`decode.py`) use the `iter_packet_bytes()` function but stopped because (a) it generates circular imports (b) it only saves a few lines of code. Do you feel strongly about this? If so, it might make more sense to put `iter_packet_bytes()` in `ccsdspy/core.py` to resolve the circular dependency issue.

ddasilva commented:
That's a reasonable comment-- I reviewed the code and most of the magic numbers are things that can be put in a file like `ccsdspy/constants.py`, eg the number of bytes in the primary header. In some cases, there are thingslike `x*256` which are effectively doing a left bitshift, and may be written better with the `<<` operator using a left shift amount defined by a constant. I have some things to clean up here.

ddasilva commented:
I cleaned up the majority of the magic numbers in the code, and introduced `ccsdspy/constants.py` the currently has `BITS_PER_BYTE` and `PRIMARY_HEADER_NUM_BYTES`. I edited the code to use these as much as possible.

Per the duplication (especially in finding the total number of bytes and APID of the packet), I created `_get_packet_total_bytes(primary_header_bytes)` and `_get_packet_apid(primary_header_bytes)`. I also expose these through utils, because they might be useful when debugging packet creation. 

Let me know if you think this is good to merge. I expect that the desire to avoid duplication and magic numbers will be an ongoing effort and not just limited to this PR.

ehsteve commented on docs/user-guide/utils.rst:
```diff
@@ -4,15 +4,83 @@
+          f"of {len(packet_bytes)} are: ")
+    print(packet_bytes[i].hex())
+
+  print()
```
I would just remove these two print statements.

ddasilva replied:
I removed these print statements

ehsteve commented on docs/user-guide/utils.rst:
```diff
@@ -4,15 +4,83 @@
+
+.. code-block::
+
+  The bytes in packet number 1 out of 10 are:
```
This statement is confusing. Sounds like you are going to print out 10 packets and this is the first. Would it be better to say "The 10 bytes in packet number 1 are:"

ddasilva replied:
I removed these print statements

ehsteve commented on docs/user-guide/utils.rst:
```diff
@@ -4,15 +4,83 @@
+====================================
+Sometimes, it is desirable to know  the number of packets in a file. For this, the `utils.count_packets()` function can be used. It's argument is a file-like object or name of a file. It accepts the optional arguments `return_missing_bytes=True` which can be used to determine the number of bytes which would be needed to be added to the file to complete the last packet (computed using the packet l ...
+
+This function works with mixed files containing multiple APIDs, which may include both fixed length and variable length packets.
```
How does it work if there are multiple APIDs? Does it give you the number of packets per APID?

ddasilva replied:
It just counts the number of packets of any APIDs. You could mix this with split_by_apid() if you wanted to count per APID. I added a comment to the doc

ehsteve commented on docs/user-guide/utils.rst:
```diff
@@ -32,3 +102,33 @@ Splitting a mixed file or stream of bytes by APID so they can be used with the `
    Writing ./apid00385.tlm...
```
everywhere else you use `.bin` for packet files.

ddasilva replied:
Yeah, but this feature isn't new so if we change it, it would break backwards incompatibility.  I changed all the other instances to `.tlm`

ehsteve commented on ccsdspy/tests/test_hs.py:
```diff
@@ -9,6 +9,7 @@
 import pytest
 
 from .. import FixedLength, VariableLength, PacketField, PacketArray
+from ..constants import BITS_PER_BYTE
```
Thanks!

ehsteve commented on ccsdspy/decode.py:
```diff
@@ -1,12 +1,92 @@
 
 
+def _get_packet_total_bytes(primary_header_bytes):
+    """Parse the  number of bytes in a packet from the bytes associated
```
```suggestion
    """Parse the number of bytes in a packet from the bytes associated
```
There are a few places throughout this code where extra spaces show up between words in text.

ehsteve commented on ccsdspy/decode.py:
```diff
@@ -1,12 +1,92 @@
 __author__ = "Daniel da Silva <mail@danieldasilva.org>"
 
 
+def _get_packet_total_bytes(primary_header_bytes):
```
Do these functions need to be private? Some of them might be helpful to users just like the utility functions. If they are marked as private then they don't show up in the documentation at all which is a shame since they have doc strings!

ddasilva replied:
They are exposed via the utils module without the private prefix in the method name.

ddasilva commented:
All changes addressed; let me know if you have any other feedback or if we're ready to merge
